File: app/src/main/java/com/wayfair/themoviedb/feature/searchresults/SearchResultsContract.java
package com.wayfair.themoviedb.feature.searchresults;

import com.pubbix.base.BaseView;
import com.wayfair.themoviedb.base.BaseRouter;
import com.wayfair.themoviedb.feature.searchresults.datamodel.SearchResultDataModel;
import com.wayfair.themoviedb.feature.searchresults.viewmodel.SearchBarViewModel;
import com.wayfair.themoviedb.feature.searchresults.viewmodel.SearchResultViewModel;

import java.util.List;

public interface SearchResultsContract {
    interface View extends BaseView {
        void addSearchBarViewModel(SearchBarViewModel viewModel);

        void addSearchResultsHeader(String searchTerms);

        void showProgressBar();

        void hideProgressBar();

        void addSearchResultBrick(SearchResultViewModel searchResultViewModel);

        void clear();
    }

    interface Router extends BaseRouter {
        void goBack();
    }

    interface Presenter {
        void setUpViews(String searchTerms);

        void fetchData(String searchTerms);

        void renderSearchResults(List<SearchResultDataModel> searchResultDataModels, String searchTerms);
    }

    interface Interactor  extends SearchBarViewModel.Interactions{
        void fetchData(String searchTerms);

        void onFetchDataSuccess(List<SearchResultDataModel> searchResultDataModels);

        void onFetchDataFailure(Throwable throwable);

        void setPresenter(Presenter presenter);
    }

    interface Repository {
        void performMultiSearch(String searchTerms);

        void setInteractor(Interactor interactor);
    }
}

File: app/src/main/java/com/wayfair/themoviedb/feature/searchresults/SearchResultsContract.kt
package com.wayfair.themoviedb.feature.searchresults

import com.wayfair.themoviedb.base.BaseRouter
import com.wayfair.themoviedb.base.BaseView
import com.wayfair.themoviedb.feature.searchresults.datamodel.SearchResultDataModel
import com.wayfair.themoviedb.feature.searchresults.viewmodel.SearchResultViewModel

interface SearchResultsContract {
    interface View : BaseView {
        fun showSearchCardView()

        fun setSearchViewText(searchTerms: String)

        fun showSearchResultsHeader(searchTerms: String)

        fun setSearchResultsHeader(searchTerms: String)

        fun showProgressBar()

        fun hideProgressBar()

        fun hideSearchViewVisibility()

        fun addSearchResultBrick(searchResultViewModel: SearchResultViewModel)
    }

    interface Router : BaseRouter

    interface Presenter {
        fun fetchData(searchTerms: String)

        fun renderSearchResults(searchResultDataModels: List<SearchResultDataModel>)
    }

    interface Interactor {
        fun fetchData(searchTerms: String)

        fun onFetchDataSuccess(searchResultDataModels: List<SearchResultDataModel>)

        fun onFetchDataFailure(throwable: Throwable)

        fun setPresenter(presenter: Presenter)
    }

    interface Repository {
        fun performMultiSearch(searchTerms: String)

        fun setInteractor(interactor: Interactor)
    }
}
